Data/DList.hs

```diff
@@ -225,6 +225,14 @@ map          :: (a -> b) -> DList a -> DList b
 map f        = foldr (cons . f) empty
 {-# INLINE map #-}
 
+-- | /O(n)/. Intersperses an element over difference lists.
+intersperse :: a -> DList a -> DList a
+intersperse sep = fromList . List.intersperse sep . toList
+
+-- | O/(n)/. Intercalate an element over nested difference lists.
+intercalate :: DList a -> [DList a] -> DList a
```

Comment:
This is rather different than the list version, whose type is:

```haskell
intercalate :: [a] -> [[a]] -> [a]
```

I believe the direct translation of types would be:

```haskell
intercalate :: DList a -> DList (DList a) -> DList a
```

Is your version useful? I'm not sure.

Reply:
The outer `DList` would have to be a `List` type since a conversion is necessary in the implementation of `intercalate` so I figured I'd just leave it as a `List` inspired off of the type signature from [`concat`](https://github.com/spl/dlist/blob/96cfa13e09975eb6f8e8d33c0b873af46b2bebdb/Data/DList.hs#L185). 

In your opinion is it still reasonable to have either of these functions at least after the type signature of [`intersperse`](https://github.com/spl/dlist/blob/f0bf6b7f5565d81d7347e0d3b07285d1f406c0ad/Data/DList.hs#L229) is revised in the spirit of [`foldr`](https://github.com/spl/dlist/blob/f0bf6b7f5565d81d7347e0d3b07285d1f406c0ad/Data/DList.hs#L219)?

Reply:
Your `intercalate` is certainly not completely unreasonable. There are are other examples that use a similar approach:

```haskell
intercalate :: ByteString -> [ByteString] -> ByteString
intercalate :: Text -> [Text] -> Text
```

And those also hint at the typical usage of `Data.List.intercalate`, which might be more concretely typed as:

```haskell
intercalate :: String -> [String] -> String
```

And I shouldn't have implied that `intercalate :: DList a -> DList (DList a) -> DList a` would be better. At this point, I'm just not sure that any version of `Data.DList.intercalate` will be useful. If you have a concrete example of where it is needed, that would be good.

Reply:
sure, [here is where i've needed it ](https://github.com/riz0id/algebraic-query/blob/c1ad97b85f00ac32dabe9002ad468c45f62ef511/src/Database/Algebraic/Backend/Internal.hs#L57)

Reply:
> In your opinion is it still reasonable to have either of these functions at least after the type signature of [`intersperse`](https://github.com/spl/dlist/blob/f0bf6b7f5565d81d7347e0d3b07285d1f406c0ad/Data/DList.hs#L229) is revised in the spirit of [`foldr`](https://github.com/spl/dlist/blob/f0bf6b7f5565d81d7347e0d3b07285d1f406c0ad/Data/DList.hs#L219)?

I'm not sure what you mean. Considering the definition from `Data.List` in `base` below, it seems to me that any `DList` version of `intersperse` is going to end up as a fold, which means the destruction (`toList`) and construction (`fromList`/`cons`) of `DList`s. If that's the case, then it would probably be better to start with a list instead of a `DList` for that scenario. If that's not the case, I will be impressed with the result!

```haskell
intersperse             :: a -> [a] -> [a]
intersperse _   []      = []
intersperse sep (x:xs)  = x : prependToAll sep xs

prependToAll            :: a -> [a] -> [a]
prependToAll _   []     = []
prependToAll sep (x:xs) = sep : x : prependToAll sep xs
```

Reply:
> sure, [here is where i've needed it ](https://github.com/riz0id/algebraic-query/blob/c1ad97b85f00ac32dabe9002ad468c45f62ef511/src/Database/Algebraic/Backend/Internal.hs#L57)

Okay, great. But what you've got is something more like:

```haskell
intercalate :: [a] -> [DList a] -> DList a
intercalate sep = Data.DList.concat . Data.List.intersperse sep
```

And, now, looking at the implementation in this PR, it seems like your type is wrong.

Reply:
Ah, you're using `OverloadedStrings`. That's why it works in your code.

Reply:
> And, now, looking at the implementation in this PR, it seems like your type is wrong.

Oh, sorry, brain glitch. It's not wrong.